### PR TITLE
Feature: Add experimental status API

### PR DESCRIPTION
**Pull request by ndelangen**
This is a proposal/implementation of a new (manager) API for storybook.

This adds the ability to assign a status to stories. Multiple statusses to be exact.
This status information can then be used in the sidebar, or addons.

The API allows addons to call: `api.setStatus({})`.
You can see tests in this PR for more in-depth examples.

I am considering adding for information, it would be nice if there was also a displayname, perhaps and some additional information the addon could except for just the raw "statuc" (which is a enum in this proposal).

Feedback is welcome.

Use-cases for this API:

The a11y addon could use this addon to set which stories have passed the check, and which ones need attention.
The coverage addon can highlight which stories fall below the coverage threshold.

The plan is that long-term there will be a "run report/testrunner over all my stories" within storybook itself.
Setting a status will be a key way to store information about individual stories from addons, that then can be consumed everywhere.

### CODE REVIEW

File: code/lib/manager-api/src/tests/stories.test.ts [hunk 1]
```diff
@@ -1629,4 +1632,81 @@ describe('stories API', () => {
...
+      expect(store.getState().status).toMatchInlineSnapshot(`
+        Object {
+          "a-story-id": Object {
+            "a-addon-id": "loading",
+          },
+          "another-story-id": Object {
+            "a-addon-id": "ready",
+          },
+        }
+      `);
```
Comment (ndelangen) [inline code/lib/manager-api/src/tests/stories.test.ts:1709]: Example usage

File: code/lib/manager-api/src/modules/stories.ts [hunk 2]
```diff
@@ -61,11 +62,14 @@ type ViewMode = 'story' | 'info' | 'settings' | string | undefined;
 type StoryUpdate = Partial<
   Pick<API_StoryEntry, 'prepared' | 'parameters' | 'initialArgs' | 'argTypes' | 'args'>
 >;
+type Status = Record<StoryId, Record<string, 'loading' | 'ready' | 'error' | 'warn' | 'unknown'>>;
```
Comment (JReinhold) [inline code/lib/manager-api/src/modules/stories.ts:65]: I feel like in general for "async" work like statuses, the status names are:

- pending (you have "loading")
- success (you have "ready"?)
- error
- warn
- unknown

that would be my recommendation.

**Review by JReinhold — COMMENTED** [review]
Looks okay, I can imagine that we'd want more meta info about the status:

```ts

type StatusUpdate = {
  status: Status;
  reporter: string; // addon name, or similar
  summary?: string;
  description?: string;
};

setStatus({
  status: "error",
  reporter: "@storybook/addon-a11y",
  summary: "The story fails 2 accessibility tests",
  description: 'The following tests failed: "missing-label", "img-missing-alt"'
})

```

**Comment by ndelangen** [comment]
When reading the data from store later, having the relationship be:

story_id => give me all statusses
story_id => give me status of addon_id

Is easier & faster when the story_id is the primary key.

I made it this structure with the sidebar needing this data later in mind.

**Comment by JReinhold** [comment]
I think @ndelangen's API suggestion makes the most sense. But given that the change is keyed by addon id, we don't necessarily need the reporter too.

File: code/lib/manager-api/src/modules/stories.ts (hunk 2 above)
Comment (shilman) [reply code/lib/manager-api/src/modules/stories.ts:65]: I like @JReinhold 's recommendation

**Review by shilman — APPROVED** [review]
Nice job & nice tests!

**Review by shilman — COMMENTED** [review]
Reading the comments, I agree with the status object `{ status, reporter, description, details }` (exact structure TBD) suggestion. I also think @tmeasday 's suggestion of `setStatus('addon-id', statusObject)` makes sense. We can store the data however we want, but that seems like a good API simplification

**Comment by tmeasday** [comment]
> I could -for sure- have the API be one way, and the actual data the other way, complicating the API code a little bit, but I'm more concerned about the confusion it would bring to addon-creators.

I would have thought addon creators needing to know to set the right `reporter` field on every update (PS it's called `context` in the GitHub status API, FWIW) would be more prone to error. 🤷 But it's not a big deal.

**Comment by ndelangen** [comment]
I'll make the change, thank you for the comments, all!

File: code/lib/manager-api/src/modules/stories.ts [hunk 3]
```diff
@@ -102,6 +114,7 @@ export interface SubAPI {
   updateStory: (storyId: StoryId, update: StoryUpdate, ref?: API_ComposedRef) => Promise<void>;
   updateDocs: (storyId: StoryId, update: DocsUpdate, ref?: API_ComposedRef) => Promise<void>;
   setPreviewInitialized: (ref?: ComposedRef) => Promise<void>;
+  experimental_updateStatus: (id: string, update: StatusUpdate) => Promise<void>;
```
Comment (tmeasday) [inline code/lib/manager-api/src/modules/stories.ts:117]: Is the `id` here the addon id? that's not obvious. I'd suggest changing the var name and considering giving it a special type.

Can we document the feature via a JSdoc comment too? I think we should get into the practice of doing that.

**Review by tmeasday — APPROVED** [review]
LGTM apart from some small quality of life improvements.

**Pull request merged**